localfs/files.go

```diff
@@ -0,0 +1,81 @@
+package localfs
+
+import (
+	"crypto/tls"
+	"io/ioutil"
+	"github.com/dbcdk/shelob/util"
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+	"github.com/dbcdk/shelob/logging"
+)
+
+var log = logging.GetInstance()
+
+func GetCerts(config *util.Config) (map[string]*tls.Certificate, error) {
+	certs := make(map[string]*tls.Certificate)
+		for name, files := range config.CertFilePairMap {
+			pubRaw, err := ioutil.ReadFile(files.PublicKey)
+			if err != nil {
+				return nil, err
+			}
+			privRaw, err := ioutil.ReadFile(files.PrivateKey)
+			if err != nil {
+				return nil, err
+			}
+			cert, err := util.ParseX509(pubRaw, privRaw)
+			if err != nil {
+				return nil, err
+			}
+			certs[name] = cert
+		}
+		return certs, nil
+}
+
+func WatchSecrets(config *util.Config, updateChan chan util.Reload) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Write == fsnotify.Write {
+					log.Info("Received inotify write event for file",
+						zap.String("file", event.Name),
+					)
+					updateChan <- util.NewReload("inotify-write-event")
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				log.Error("Inotify watch error",
+					zap.String("error", err.Error()),
+				)
+			case stopping := <-config.State.ShutdownChan:
+				if stopping {
+					log.Info("Stopping inotify watch loop")
+					watcher.Close()
+					return
+				}
+			}
+		}
+	}()
+
+	for _, pair := range config.CertFilePairMap {
```

Comment:
I'm not sure I understand how this works in the presence of reconciliation on a timer. Can you explain what happens if we're here before the files have been provisioned and `return err` for that reason? Who calls this again to ensure watches are set up later?

Reply:
Added ac2b7bd.

It removes the early "forking" in certs.go and moves the responsibility of "go"-spawning to the kubernetes and localfs-impls respectively. This makes it possible to synchronously return an error to "main" during the entire watch-setup-process right until the actual watch process is started.

Note: This doesn't handle inotify errors or deleted secrets. In these cases, the watch loop will die and the watch is never attempted again. Let me know if you think I should fix this as well.

Reply:
OK, now we'll definitely die if we don't find everything, allowing restarts of the process until (hopefully) files appear. I gets it.